services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadPartRequestPublisher.java

```diff
@@ -0,0 +1,379 @@
+                        return String.format("Starting to prepare requests for parts %d-%d. Remaining"
+                                             + " parts: %d",
+                                             index,
+                                             index + numOfParts - 1, finalRemainingParts);
+                    });
+
+                    generateRequests(remainingParts, numOfParts);
+                } finally {
+                    isPreparingRequests.set(false);
+                }
+            }
+        }
+
+        private void generateRequests(int remainingParts, int numOfParts) {
+            for (int i = 0; i < numOfParts; i++) {
+                log.trace(() -> String.format("Processing current part %d", numOfRemainingParts.get()));
+
+                long currentPartSize = remainingParts == 1 ? lastPartSize : optimalPartSize;
+
+                int numOfBuffersForCurrentPart = (int) Math.ceil(currentPartSize / (double) DEFAULT_BUFFER_SIZE);
+
+                SimplePublisher<ByteBuffer> simplePublisher = new SimplePublisher<>();
+
+                sendByteBuffersToSimplePublisher(numOfBuffersForCurrentPart, currentPartSize, simplePublisher);
+
+                int currentPartNum = partNumberIndex.getAndIncrement();
+                UploadPartRequest uploadPartRequest =
+                    RequestConversionUtils.toUploadPartRequest(putObjectRequest,
+                                                               currentPartNum,
+                                                               uploadId);
+
+                remainingParts = numOfRemainingParts.decrementAndGet();
+                requests.add(Pair.of(uploadPartRequest, new AsyncRequestBody() {
+                    @Override
+                    public Optional<Long> contentLength() {
+                        return Optional.of(currentPartSize);
+                    }
+
+                    @Override
+                    public void subscribe(Subscriber<? super ByteBuffer> s) {
+                        simplePublisher.subscribe(s);
+                    }
+                }));
+
+                if (remainingParts == 0) {
+                    log.debug(() -> "AsyncRequestBody has been drained. All requests have been generated");
+                    upstreamDone = true;
+                }
+            }
+        }
+
+        private void sendByteBuffersToSimplePublisher(int numOfBuffersForCurrentPart,
+                                                      long remainingBytes,
+                                                      SimplePublisher<ByteBuffer> simplePublisher) {
+            for (int bufferIndex = 0; bufferIndex < numOfBuffersForCurrentPart; bufferIndex++) {
+                int bufferSize = adjustBufferSize(numOfBuffersForCurrentPart, remainingBytes, bufferIndex);
+
+                ByteBuffer byteBuffer = ByteBuffer.allocate(bufferSize);
+                byteBufferStoringSubscriber.blockingTransferTo(byteBuffer);
```

Comment:
I thought about it. I used a different thread pool to buffer data in advance so that we can parallelize upload later. If we use the same threads and we read data in serial, the upload requests would be sent in serial as well. LMK if you can think of a way that avoids this and achieves parallelization.